consensus/XDPoS/engines/engine_v2/utils.go

```diff
@@ -108,6 +95,53 @@ func (x *XDPoS_v2) signSignature(signingHash common.Hash) (types.Signature, erro
 	return signedHash, nil
 }
 
+func (x *XDPoS_v2) countValidSignatures(messageHash common.Hash, signatures []types.Signature, candidates []common.Address) (int, error) {
+	signatureList := make([]types.Signature, len(signatures))
+	pubkeys := make([]common.Address, len(signatures))
+
+	eg, ctx := errgroup.WithContext(context.Background())
+	eg.SetLimit(runtime.NumCPU())
+	for i, signature := range signatures {
+		eg.Go(func() error {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			default:
+				verified, signerAddress, err := x.verifyMsgSignature(messageHash, signature, candidates)
+				if err != nil {
+					log.Error("[verifySignatures] Error while verfying QC message signatures", "error", err)
+					return err
+				}
+				if !verified {
+					return fmt.Errorf("signature verification failed, signer is not part of masternode list. Signature: %v, SignedMessage: %v, SignerAddress: %v, Masternodes: %v", signature, messageHash.Hex(), signerAddress, candidates)
+				}
+				signatureList[i] = signature
+				pubkeys[i] = signerAddress
+				
+				return nil
+			}
+		})
+	}
+	err := eg.Wait()
+	if err != nil {
+		return 0, err
+	}
+
+	// check uniqueness
+	keys := make(map[common.Address]struct{}, len(signatureList))
+	for i, _ := range signatureList {
+		pubkeyHex := pubkeys[i]
+		if _, ok := keys[pubkeyHex]; !ok {
+			keys[pubkeyHex] = struct{}{}
+		} else {
+			log.Warn("[verifySignatures] duplicate signing found", "pubkey", pubkeyHex, "signedMessage", messageHash.Hex(), "signature", signatureList[i])
```

Comment:
should use pubkeyHex.Hex() in log message

Reply:
No, log use TerminalStringer:

```go
// TerminalStringer is an analogous interface to the stdlib stringer, allowing
// own types to have custom shortened serialization formats when printed to the
// screen.
type TerminalStringer interface {
	TerminalString() string
}
```

type Hash implement TerminalStringer:

```go
// TerminalString implements log.TerminalStringer, formatting a string for console
// output during logging.
func (h Hash) TerminalString() string {
	return fmt.Sprintf("%x..%x", h[:3], h[29:])
}
```

But type Address does not implement TerminalStringer.

Reply:
Sorry, my mistake. `log` prefer to use `TerminalStringer`, `fmt. Stringer` is used if `TerminalStringer` is not implemented.

```go
	case TerminalStringer:
		return appendEscapeString(tmp, v.TerminalString())
	case fmt.Stringer:
		return appendEscapeString(tmp, v.String())
```